build.gradle

```diff
@@ -48,6 +50,20 @@ ext {
 	}
 	println "JDK Javadoc link for this build is ${rootProject.jdkJavadoc}"
 
+	versionNumber = VersionNumber.parse(version.toString())
+	if (versionNumber.qualifier == null || versionNumber.qualifier.size() == 0) {
+		osgiVersion = "${version}.RELEASE"
+		println "$version is a release, will use $osgiVersion for bnd"
+	}
+	else if (versionNumber.qualifier.equalsIgnoreCase("SNAPSHOT")) {
+		osgiVersion = "${versionNumber.major}.${versionNumber.minor}.${versionNumber.micro}.BUILD-SNAPSHOT"
```

Comment:
For export packages, the plugin removes the qualifier from the version when it is `*-SNAPSHOT` and thus we have only
`${versionNumber.major}.${versionNumber.minor}.${versionNumber.micro}` which is not correct for snapshots versions.
Better option here will be to add `BUILD-timestamp` thus we will be able to distinguish the different snapshot versions.

Reply:
There is an BND instruction that tells BND what to do when an "-SNAPSHOT" qualifier is encountered in the bundle version: https://bnd.bndtools.org/instructions/snapshot.html

One possibility would be to use:
-snapshot: BUILD-${tstamp}

Which would change:
3.4.1-SNAPSHOT
to
3.4.1.202011251350

Reply:
Yes but then when you have two bundles with `3.4.1.202011251350` and `3.4.1` and import package [3.4, 3.5) thew first will be preferred correct? While we want the bundles produced by the release to be preferred in comparison to the SNAPSHOT builds.

So the snapshots will be `3.4.1.BUILD-202011251350` while the release will be `3.4.1.RELEASE`.
Is that correct?

Reply:
@violetagg you are right that when the major/minor and micro version matches in OSGi the qualifier is then compared lexicographically. Also a version without a qualifier is always ranked lower than a version with a qualifier e.g. 3.4.1 will be lower than 3.4.1.RELEASE.

So with all the examples we get following order:
- 3.4.1 (lowest)
- 3.4.1.202011251350
- 3.4.1.BUILD-202011251350
- 3.4.1.RELEASE (highest)

Thats the version compareTo method of bnd:
```java
public int compareTo(Version other) {
	if (other == this)
		return 0;

	Version o = other;
	int cmp = major - o.major;
	if (cmp != 0)
		return cmp;

	cmp = minor - o.minor;
	if (cmp != 0)
		return cmp;

	cmp = micro - o.micro;
	if (cmp != 0)
		return cmp;

	if (qualifier != null)
		cmp = 1;
	if (o.qualifier != null)
		cmp += 2;

	switch (cmp) {
		case 0 :
			return 0;
		case 1 :
			return 1;
		case 2 :
			return -1;
	}
	return qualifier.compareTo(o.qualifier);
}
```

There are probably more ways to solve this but I also agree that using a timestamp for snapshots would be better.

Reply:
🤕👿  now I'm not changing the `Bundle-Version` for `-SNAPSHOT` but instead using the bnd instruction `"-snapshot" : 'BUILD-${tstamp}'` and what I get is this:

```yaml
Bundle-Version: 3.4.1.BUILD-202011251646
Created-By: 14.0.2 (AdoptOpenJDK)
Export-Package: reactor.adapter;uses:="org.reactivestreams,reactor.cor
 e.publisher";version="3.4.1.SNAPSHOT"
```
oh what a joy!

Reply:
Oh wow... I haven't use that instruction before but I wouldn't say that is expected behavior... in the meantime the currently proposed solution might be the best.

Reply:
@swimmesberger do you mean the code currently committed in this PR branch? ie the one that produces `.BUILD-SNAPSHOT` for snapshots and `.RELEASE` for releases? (just to be sure)

Reply:
@simonbasle  I'm currently trying to reproduce the -snapshot instruction issue. I will report back when I have come to a conclusion.

Reply:
@simonbasle although the -snapshot instruction is not working (see my last comment for reasoning) we should still use BUILD-${tstamp} here with goovy methods e.g. something along the lines:

```groovy
def sdf = new SimpleDateFormat("yyyyMMddHHmm");
sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
def tstamp = sdf.format(new Date())
```

Reply:
What about this

```
diff --git a/build.gradle b/build.gradle
index 5cc2d77e..170d0244 100644
--- a/build.gradle
+++ b/build.gradle
@@ -1,4 +1,5 @@
 import org.gradle.util.VersionNumber
+import java.text.SimpleDateFormat
 
 /*
  * Copyright (c) 2011-2017 Pivotal Software Inc, All Rights Reserved.
@@ -42,6 +43,12 @@ apply plugin: "io.reactor.gradle.detect-ci"
 apply from: "gradle/asciidoc.gradle" //asciidoc (which is generated from root dir)
 apply from: "gradle/releaser.gradle"
 
+def buildTime() {
+	def sdf = new SimpleDateFormat("yyyyMMddHHmm");
+	sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
+	return sdf.format(new Date())
+}
+
 ext {
 	jdk = JavaVersion.current().majorVersion
 	jdkJavadoc = "https://docs.oracle.com/javase/$jdk/docs/api/"
@@ -56,7 +63,7 @@ ext {
 		println "$version is a release, will use $osgiVersion for bnd"
 	}
 	else if (versionNumber.qualifier.equalsIgnoreCase("SNAPSHOT")) {
-		osgiVersion = "${versionNumber.major}.${versionNumber.minor}.${versionNumber.micro}.BUILD-SNAPSHOT"
+		osgiVersion = "${versionNumber.major}.${versionNumber.minor}.${versionNumber.micro}.BUILD-${buildTime()}"
 		println "$version is a snapshot, will use $osgiVersion for bnd"
 	}
 	else {
diff --git a/reactor-core/build.gradle b/reactor-core/build.gradle
index 4f0d63ef..97aa6502 100644
--- a/reactor-core/build.gradle
+++ b/reactor-core/build.gradle
@@ -38,7 +38,7 @@ ext {
 		"Export-Package": [
 			"!*internal*",
 			"!reactor.blockhound*",
-			"reactor.*"
+			"reactor.*;version=$osgiVersion"
 		].join(","),
 		"Import-Package": [
 			"!javax.annotation",
diff --git a/reactor-test/build.gradle b/reactor-test/build.gradle
index 116a688f..3e39d2f7 100644
--- a/reactor-test/build.gradle
+++ b/reactor-test/build.gradle
@@ -24,7 +24,7 @@ ext {
 	bndOptions = [
 		"Export-Package": [
 			"!*internal*",
-			"reactor.test*;-noimport:=true"
+			"reactor.test*;version=$osgiVersion;-noimport:=true"
 		].join(","),
 		"Import-Package": [
 			"!javax.annotation",
```

Reply:
@violetagg That would work and will result in "3.4.1.BUILD-202011251646" versions for exported packages but I don't think that it is strictly required? 🤔 Using the default behavior would result in:
Bundle-Version: 3.4.1.BUILD-202011251646
Export-Package: xyz;version=3.4.1

Assuming OSGi uses the same version order for packages it would rank release packages before snapshot packages which would be good. As I have said if there are multiple bundles which provide the same package with the same version maybe the bundle version is then used for ordering that would bean the package version qualifier wouldn't be strictly required but I'm not sure.

For now it's probably the best solution 👌

Reply:
According to the spec:
https://docs.osgi.org/specification/osgi.core/7.0.0/service.resolver.html#i3344409

or resolving an OSGi framework the specifications outlines a number of heuristics that guide the order of wiring bundles and packages:
- A resource that is already resolved, that is, it is already wired
- The highest version
- The lowest bundle id

Therefore I would assume that the same version order used for Bundle-Version is also used for the package version 🙂

Reply:
> @violetagg That would work and will result in "3.4.1.BUILD-202011251646" versions for exported packages but I don't think that it is strictly required? 🤔 Using the default behavior (without specifying a export-package version) would result in:
> Bundle-Version: 3.4.1.BUILD-202011251646
> Export-Package: xyz;version=3.4.1
> 
> Assuming OSGi uses the same version order for packages it would rank release packages before snapshot packages which would be good. If there are multiple bundles which provide the same package with the same version maybe the bundle version is then used for ordering that would mean the package version qualifier wouldn't be strictly required but I'm not sure.
> 
> For now it's probably the best solution 👌

Let's assume that you have two bundles with the following versions `Bundle-Version: 3.4.1.BUILD-1` and the second `Bundle-Version: 3.4.1.BUILD-2`. The first one has an issue and the second one no. You want to test both of them in order to find the issue.
Both bundles are installed and your bundle want to test both of them. You can do it very easy when specifying the concrete export-package version when you importing the package.
You do not need to install/uninstall bundles, you do not need to require a bundle.

Is that a valid scenario or if you think that you are just fine with the current solution we can do it like this.

Reply:
@violetagg you are right that might be a valid use-case and the changes required are not that big. In many OSGi specific projects packages are versioned completely separately than the bundle version but for projects like this one it might be a good idea to include the qualifier for the exported packages.